alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Bbody.hs

```diff
@@ -0,0 +1,207 @@
+  | TooManyExUnits
+      !ExUnits
+      -- ^ Computed Sum of ExUnits for all plutus scripts
+      !ExUnits
+      -- ^ Maximum allowed by protocal parameters
+  deriving (Generic)
+
+deriving instance
+  (Era era, Show (PredicateFailure (Core.EraRule "LEDGERS" era))) =>
+  Show (AlonzoBbodyPredFail era)
+
+deriving instance
+  (Era era, Eq (PredicateFailure (Core.EraRule "LEDGERS" era))) =>
+  Eq (AlonzoBbodyPredFail era)
+
+deriving anyclass instance
+  (Era era, NoThunks (PredicateFailure (Core.EraRule "LEDGERS" era))) =>
+  NoThunks (AlonzoBbodyPredFail era)
+
+-- TODO  Do we need CBOR instances? he shelley (BbodyPredicateFailure era) doesn't seem to have one.
+
+-- ========================================
+-- The STS instance
+
+-- | The uninhabited type that marks the STS Alonzo Era instance.
+data AlonzoBBODY era
+
+bbodyTransition ::
+  forall (someBBODY :: Type -> Type) era.
+  ( -- Conditions that the Abstract someBBODY must meet
+    STS (someBBODY era),
+    Signal (someBBODY era) ~ Block era,
+    PredicateFailure (someBBODY era) ~ AlonzoBbodyPredFail era,
+    BaseM (someBBODY era) ~ ShelleyBase,
+    State (someBBODY era) ~ BbodyState era,
+    Environment (someBBODY era) ~ BbodyEnv era,
+    -- Conditions to be an instance of STS
+    Embed (Core.EraRule "LEDGERS" era) (someBBODY era),
+    Environment (Core.EraRule "LEDGERS" era) ~ LedgersEnv era,
+    State (Core.EraRule "LEDGERS" era) ~ LedgerState era,
+    Signal (Core.EraRule "LEDGERS" era) ~ Seq (Core.Tx era),
+    -- Conditions to define the rule in this Era
+    BlockDecoding era,
+    HasField "_d" (Core.PParams era) UnitInterval,
+    HasField "_maxBlockExUnits" (Core.PParams era) ExUnits,
+    HasField "totExunits" (Core.Tx era) ExUnits,
+    Era era -- supplies WellFormed HasField, and Crypto constraints
+  ) =>
+  TransitionRule (someBBODY era)
+bbodyTransition =
+  judgmentContext
+    >>= \( TRC
+             ( BbodyEnv pp account,
+               BbodyState ls b,
+               Block (BHeader bhb _) txsSeq
+               )
+           ) -> do
+        let txs = txSeqTxns txsSeq
+            actualBodySize = bBodySize txsSeq
+            actualBodyHash = bbHash txsSeq
+
+        actualBodySize == fromIntegral (hBbsize bhb)
+          ?! (ShelleyInAlonzoPredFail $ WrongBlockBodySizeBBODY actualBodySize (fromIntegral $ hBbsize bhb))
+
+        actualBodyHash == bhash bhb
+          ?! (ShelleyInAlonzoPredFail $ InvalidBodyHashBBODY @era actualBodyHash (bhash bhb))
+
+        ls' <-
+          trans @(Core.EraRule "LEDGERS" era) $
+            TRC (LedgersEnv (bheaderSlotNo bhb) pp account, ls, StrictSeq.fromStrict txs)
+
+        -- Note that this may not actually be a stake pool - it could be a genesis key
+        -- delegate. However, this would only entail an overhead of 7 counts, and it's
+        -- easier than differentiating here.
+        let hkAsStakePool = coerceKeyRole . issuerIDfromBHBody $ bhb
```

Comment:
the coercion should happen inside incrBlocks

Reply:
Left a TODO comment about why this might be confusing.